### ISSUE TITLE
Static type checker crashed on syntax error in lib/曆法

### ISSUE THREAD

**Issue by statementreply** [LingDong-/wenyan-lang#495]
https://github.com/LingDong-/wenyan-lang/blob/90eb9a1f787ba4a69ad7c5daf6a561a9ec7dba55/lib/%E6%9B%86%E6%B3%95.wy#L152
`也` is invalid here. The static type checker crashed on it.

```
Uncaught TypeError: Cannot read property '言年月日' of undefined
    at checkscopei (typecheck.js:89)
    at checkscopethis (typecheck.js:130)
    at typecheck (typecheck.js:270)
    at compile (parser.js:736)
    at compile (parser.js:766)
```

Minimum repro

```
吾有一術。名之曰「甲」。是術曰。
    乃得一也。
是謂「甲」之術也。
```

## Checklist

- [x] If this is a feature request, it's NOT listed [Here](https://github.com/LingDong-/wenyan-lang#feature-requests)
- [x] If this is a bug report, it's NOT listed [Here](https://github.com/LingDong-/wenyan-lang#known-bugs)

**Comment by LingDong-** [comment]
@statementreply Thanks for spotting the issue!
It is now fixed: 55d288581783c183b0a40caeb662bf6757302022
It gives a proper error message instead of crashing.

The problem is due to the fact that in the typechecker, variables and types are stored on a stack of scopes. every time end-block statement (e.g. 也 云云) is found, a scope on the stack of scopes is popped. When there are too many end-blocks (such as in this case), scope index gets negative and invokes JavaScript's signature `undefined`.

A better place to check for this error is probably in `token2asc`, but since currently it doesn't keep track of scopes, I put it temporarily in the typechecker. Indeed, one of my next steps is to improve `token2asc` to catch more syntax errors like these.

I also fixed another issue regarding 曆法, which is since 取底除 is an imported function, its return type is not known and cause the checker to deny the fields it actually contains.

曆法 should compile without errors now, once the extraneous 也 is removed.

Thanks!

**Comment by statementreply** [comment]
A related issue: there are several missing end-block statements in 曆法, but they are detected by neither `tokene2asc` nor `typecheck`. I believe we could also temporarily assert `scope.length === 0` at the end of `typecheck` to help detecting such kind of issues.

I'll fix the bugs in 曆法 later tonight (UTC+8).
